src/jit/block.cpp

```diff
@@ -1374,6 +1374,56 @@ BasicBlock* Compiler::bbNewBasicBlock(BBjumpKinds jumpKind)
     return block;
 }
 
+//------------------------------------------------------------------------
+// blockhasEHBoundaryIn: Determine if this block begins at an EH boundary.
+//
+// Return Value:
+//    True iff the block is the target of an EH edge; false otherwise.
+//
+bool BasicBlock::hasEHBoundaryIn()
+{
+#if FEATURE_EH_FUNCLETS
+    return ((bbCatchTyp != BBCT_NONE) || (bbFlags & BBF_FUNCLET_BEG));
+#else  // !FEATURE_EH_FUNCLETS
+    return (bbCatchTyp != BBCT_NONE);
+#endif // !FEATURE_EH_FUNCLETS
+}
+
+//------------------------------------------------------------------------
+// blockhasEHBoundaryOut: Determine if this block ends in an EH boundary.
+//
+// Return Value:
+//    True iff the block ends in an exception boundary that requires that all lclVars
+//    have an up-to-date value on the stack; false otherwise.
+//
+bool BasicBlock::hasEHBoundaryOut()
```

Comment:
I don't think this is correct. I think you need to use something like:
```
return (bbNext == nullptr) || !BasicBlock::sameEHRegion(this, bbNext))
```

The problem is there could be multiple `BBJ_CATCHRET` in a single handler. And there's no need for them to be at the end of the handler.

With `BBJ_KEEP_BBJ_ALWAYS`, I also don't think there is a requirement for them to be at the end of a handler.

With EHFILTERRET/EHFINALLYRET, I do believe the incoming IL requires them to be at the end of the handlers, and we probably don't alter that, but I also don't think there is anything preventing us from altering that by, for example, reordering the blocks in the handler and putting things at the end, after these blocks.

Reply:
I guess the only issue here is the lexical order versus flowgraph order. Maybe you need something like:
```
for each succ in successors:
{
  if (!BasicBlock::sameEHRegion(this, succ)) return true;
}
return false;
```

Reply:
OK, now I'm realizing that this method is probably the wrong way to go about this. Knowing where the "in" boundaries are is important because those blocks must re-load any lclVars that are used. However, the "out" boundaries are different, because we will always ensure that the EH vars have valid values on the stack. These out-edges matter because:
- there are places where we expect the set of live vars in registers to be empty. For those we must not have live lclVars in registers.
- if **all** successors are possible EH targets, then it isn't **useful** to keep lclVars in registers

I'm going to have to reconsider how to capture this.